src/basic/fs-util.c

```diff
@@ -1100,7 +1100,9 @@ int xopenat(int dir_fd, const char *path, int open_flags, XOpenFlags xopen_flags
         assert(dir_fd >= 0 || dir_fd == AT_FDCWD);
 
         if (isempty(path)) {
-                assert(!FLAGS_SET(open_flags, O_CREAT|O_EXCL));
+                /* assert_se() is used to avoid build failure due to false positive GCC _FORTIFY_SOURCE error
+                 * see issue #29408 */
+                assert_se(!FLAGS_SET(open_flags, O_CREAT|O_EXCL));
```

Comment:
> if NDEBUG is set this is precisely a kind of test that should be suppressed.

I agree. Which is why I suggested a `ASSUME()` macro in the bug report as a potential solution.

Greping through the source code, I see that systemd already has such a macro called `assert_not_reached` which expands to `__builtin_unreachable()` when `NDEBUG` is set.

This should elide away the check from the generated asm but still give GCC information to avoid the FP. Do you find this to be an acceptable solution?

Reply:
No need for a separate macro. Since we afaik already override the glibc assert with our own, we might as well just replace the NDEBUG version with `__builtin_assume`, no?

Also, that `assert_not_reached` definition is for `SD_BOOT` only right now. The userspace one always expands to `log_assert_failed_unreachable()`.